relayer/src/chain/cosmos/retry.rs

```diff
@@ -0,0 +1,186 @@
+use core::future::Future;
+use core::pin::Pin;
+use core::time::Duration;
+use ibc_proto::google::protobuf::Any;
+use std::thread;
+use tendermint::abci::Code;
+use tendermint_rpc::endpoint::broadcast::tx_sync::Response;
+use tendermint_rpc::HttpClient;
+use tonic::codegen::http::Uri;
+use tracing::{debug, error, warn};
+
+use crate::chain::cosmos::query::account::refresh_account;
+use crate::chain::cosmos::tx::estimate_fee_and_send_tx;
+use crate::chain::cosmos::types::account::Account;
+use crate::config::types::Memo;
+use crate::config::ChainConfig;
+use crate::error::Error;
+use crate::keyring::KeyEntry;
+use crate::sdk_error::sdk_error_from_tx_sync_error_code;
+
+// Maximum number of retries for send_tx in the case of
+// an account sequence mismatch at broadcast step.
+const MAX_ACCOUNT_SEQUENCE_RETRY: u64 = 1;
+
+// Backoff multiplier to apply while retrying in the case
+// of account sequence mismatch.
+const BACKOFF_MULTIPLIER_ACCOUNT_SEQUENCE_RETRY: u64 = 300;
+
+// The error "incorrect account sequence" is defined as the unique error code 32 in cosmos-sdk:
+// https://github.com/cosmos/cosmos-sdk/blob/v0.44.0/types/errors/errors.go#L115-L117
+const INCORRECT_ACCOUNT_SEQUENCE_ERR: u32 = 32;
+
+/// Try to `send_tx` with retry on account sequence error.
+/// An account sequence error can occur if the account sequence that
+/// the relayer caches becomes outdated. This may happen if the relayer
+/// wallet is used concurrently elsewhere, or when tx fees are mis-configured
+/// leading to transactions hanging in the mempool.
+///
+/// Account sequence mismatch error can occur at two separate steps:
+///   1. as Err variant, propagated from the `estimate_gas` step.
+///   2. as an Ok variant, with an Code::Err response, propagated from
+///     the `broadcast_tx_sync` step.
+///
+/// We treat both cases by re-fetching the account sequence number
+/// from the full node.
+/// Upon case #1, we do not retry submitting the same tx (retry happens
+/// nonetheless at the worker `step` level). Upon case #2, we retry
+/// submitting the same transaction.
+pub async fn send_tx_with_account_sequence_retry(
+    config: &ChainConfig,
+    rpc_client: &HttpClient,
+    grpc_address: &Uri,
+    key_entry: &KeyEntry,
+    account: &mut Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
+    retry_counter: u64,
+) -> Result<Response, Error> {
+    do_send_tx_with_account_sequence_retry(
+        config,
+        rpc_client,
+        grpc_address,
+        key_entry,
+        account,
+        tx_memo,
+        messages,
+        retry_counter,
+    )
+    .await
+}
+
+fn do_send_tx_with_account_sequence_retry<'a>(
+    config: &'a ChainConfig,
+    rpc_client: &'a HttpClient,
+    grpc_address: &'a Uri,
+    key_entry: &'a KeyEntry,
+    account: &'a mut Account,
+    tx_memo: &'a Memo,
+    messages: Vec<Any>,
+    retry_counter: u64,
+) -> Pin<Box<dyn Future<Output = Result<Response, Error>> + 'a>> {
```

Comment:
Do we need this explicit result type instead of an `async fn` because the compiler cannot infer the proper lifetime bound?

Reply:
Yeah Rust do not currently support recursive async functions with the `async fn` sugar. I have added some comments to explain it.

Reply:
Right, didn't realize it was recursive.

Thanks for the comments. Since afaik this is the only occurrence of such a recursive async fn I don't we need to use https://docs.rs/async-recursion/latest/async_recursion/ but I am leaving it here just in case.

Reply:
Yeah probably don't need that for now. Actually aside from the boilerplate, using `Pin<Box<dyn Future>>` is not that difficult or intimidating once you wrote it a few times.